### PR TITLE
StatsD Module

### PR DESCRIPTION
**Pull request by mamercad**
##### SUMMARY

Adding support to community.general for StatsD metrics.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

community.general.statsd

##### ADDITIONAL INFORMATION

Locally, I'm developing against the [StatsD Exporter](https://github.com/prometheus/statsd_exporter) (it uses the same line protocol as StatsD/Graphite proper).

```paste below
❯ cat ~/Desktop/dev/ansible/test.yml
---
- hosts: localhost
  connection: local
  gather_facts: no

  tasks:

    - name: Increment the metric my_counter by 1
      community.general.statsd:
        host: localhost
        port: 9125
        protocol: tcp
        metric: my_counter
        metric_type: counter
        value: 1

    - name: Set the gauge my_gauge to 7
      community.general.statsd:
        host: localhost
        port: 9125
        protocol: tcp
        metric: my_gauge
        metric_type: gauge
        value: 7

❯ ansible-playbook -i localhost, -c local ~/Desktop/dev/ansible/test.yml -v
PLAY [localhost] *******************************************************************************************************************************************************************************************************************************************

TASK [Increment the metric my_counter by 1] ****************************************************************************************************************************************************************************************************************
Thursday 11 February 2021  14:33:25 -0500 (0:00:00.061)       0:00:00.061 ***** 
changed: [localhost] => changed=true 
  msg: Sent counter my_counter -> 1 to StatsD

TASK [Set the gauge my_gauge to 7] *************************************************************************************************************************************************************************************************************************
Thursday 11 February 2021  14:33:25 -0500 (0:00:00.458)       0:00:00.520 ***** 
changed: [localhost] => changed=true 
  msg: Sent gauge my_gauge (delta=False) -> 7 to StatsD

PLAY RECAP *************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

❯ curl -s localhost:9102/metrics | grep my_
# HELP my_counter Metric autogenerated by statsd_exporter.
# TYPE my_counter counter
my_counter 1
# HELP my_gauge Metric autogenerated by statsd_exporter.
# TYPE my_gauge gauge
my_gauge 7
```

### CODE REVIEW

File: plugins/modules/monitoring/statsd.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright: Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from ansible.module_utils.basic import AnsibleModule
```
Comment (russoz) [inline plugins/modules/monitoring/statsd.py:6]: See other modules for where and how the imports should be placed in the module file.

**Comment by mamercad** [comment]
@russoz Thanks for the help!

File: plugins/modules/monitoring/statsd.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+# -*- coding: utf-8 -*-
+# Copyright: Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+module: statsd
+short_description: Send metrics to StatsD
```
Comment (felixfontein) [inline plugins/modules/monitoring/statsd.py:11]:
```suggestion
short_description: Send metrics to StatsD
version_added: 2.1.0
```

File: plugins/modules/monitoring/statsd.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+description:
+  - The C(statsd) module sends metrics to StatsD.
+  - For more information, see U(https://statsd-metrics.readthedocs.io/en/latest/).
+  - Supported metric types are C(counter) and C(gauge).
+  - Currently unupported metric types are C(timer), C(set), and C(gaugedelta).
```
Comment (felixfontein) [inline plugins/modules/monitoring/statsd.py:16]: I would put that into the same paragraph:
```suggestion
  - Supported metric types are C(counter) and C(gauge).
    Currently unupported metric types are C(timer), C(set), and C(gaugedelta).
```

File: plugins/modules/monitoring/statsd.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    host: localhost
+    port: 9125
+    protocol: tcp
+    metric: my_gauge
+    metric_type: gauge
+    value: 7
+'''
+
+
+from statsd import StatsClient, TCPStatsClient
```
Comment (felixfontein) [inline plugins/modules/monitoring/statsd.py:96]: You need to use import guards and handle the failed `ImportError` in the main module code (search for `missing_required_lib` in existing modules).

File: tests/unit/plugins/modules/monitoring/test_statsd.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+from ansible_collections.community.general.plugins.modules.monitoring import statsd
+from ansible_collections.community.general.tests.unit.plugins.modules.utils import AnsibleExitJson, AnsibleFailJson, ModuleTestCase, set_module_args
+
+
+class TestStatsDModule(ModuleTestCase):
+
+    def setUp(self):
+        super(TestStatsDModule, self).setUp()
+        self.module = statsd
```
Comment (felixfontein) [inline tests/unit/plugins/modules/monitoring/test_statsd.py:15]: The tests need to do a little bit more than this.

Also, I'd suggest you do not add `statsd` to the test requirements, but instead mock the parts of that library you are using. That allows to ensure that the module passes on the expected information to the library, and avoids having to deal with internals of the `statsd` library in the unit tests.

Comment (mamercad) [reply tests/unit/plugins/modules/monitoring/test_statsd.py:15]: Thanks for helping, I'm working on it.

File: plugins/modules/monitoring/statsd.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+      - The prefix to add to the metric.
+  value:
+    type: int
+    required: true
+    description:
+      - The value of the metric.
+  delta:
+    type: bool
+    default: False
+    choices: [False, True]
```
Comment (felixfontein) [inline plugins/modules/monitoring/statsd.py:72]:
```suggestion
```
This isn't needed.

File: plugins/modules/monitoring/statsd.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            state=dict(type='str', default='present', choices=['present']),
+            host=dict(type='str', default='localhost'),
+            port=dict(type='int', default=8125),
+            protocol=dict(type='str', default='udp', choices=['udp', 'tcp']),
+            timeout=dict(type='float', default=1.0),
+            metric=dict(type='str', required=True),
+            metric_type=dict(type='str', required=True, choices=['counter', 'gauge']),
+            metric_prefix=dict(type='str', default=''),
+            value=dict(type='int', required=True),
+            delta=dict(type='bool', default=False, choices=[False, True]),
```
Comment (felixfontein) [inline plugins/modules/monitoring/statsd.py:128]:
```suggestion
            delta=dict(type='bool', default=False),
```

**Review by felixfontein — COMMENTED** [review]
Everything else LGTM!

File: plugins/modules/monitoring/statsd.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    description:
+      - The prefix to add to the metric.
+  value:
+    type: int
+    required: true
+    description:
+      - The value of the metric.
+  delta:
+    type: bool
+    default: False
```
Comment (felixfontein) [inline plugins/modules/monitoring/statsd.py:71]:
```suggestion
    default: false
```

**Comment by felixfontein** [comment]
@mamercad I've just commited my last tiny suggestions so I can merge this for today's 2.1.0 release. I hope that's fine for you :)

**Comment by felixfontein** [comment]
@mamercad thanks for contributing this!
@russoz thanks for reviewing!

**Comment by mamercad** [comment]
> @mamercad I've just commited my last tiny suggestions so I can merge this for today's 2.1.0 release. I hope that's fine for you :)

Ah, thanks very much. I have no specific timeframe, thanks for helping along the way. I've been doing `shell: echo "foo:1|c" | nc /dev/tcp/my.statsd.host/9125` as a workaround in my playbooks.

**Pull request merged**